### PR TITLE
fix(sdk): warp apply fixes for ISM initialization, routing updates, and nonce conflicts

### PR DESCRIPTION
**Pull request by troykessler**
### Description

Three fixes to make `warp apply` work correctly for ownership migrations:

1. **`HyperlaneIsmFactory`**: Added `isInitialized` guard before calling `initialize()` on `FALLBACK_ROUTING` ISMs (both standard and ZkSync paths). On retry, the contract was already initialized, causing a revert.

2. **`EvmIsmModule`**: Added `tryUpdateContainerIsm` to update `AGGREGATION` and `AMOUNT_ROUTING` ISM sub-modules in-place when only owners change, avoiding a full redeploy of the ISM tree.

3. **`multisend`**: Replaced static `SAFE_NONCE_OVERRIDES` map with automatic `getNextNonce()` lookup via the Safe API, which accounts for pending transactions and avoids nonce conflicts.

### Backward compatibility

Yes

### Testing

Manual — ran `warp apply` for USDC/moonpay and USDT/moonpay on BSC end-to-end.

### CODE REVIEW

File: typescript/sdk/src/ism/EvmIsmModule.ts
Comment (paulbalaji) [inline typescript/sdk/src/ism/EvmIsmModule.ts:1]: Feels like an extraneous optimisation 

Would prefer to focus the PRS on bsc changes and move sdk changes to a separate or for reviewing independently

**Review by paulbalaji — COMMENTED** [review]
BSC/moonpay config fix looks clean and is ready to go: `awIcas.bsc` restored, USDC + USDT Citrea/Moonpay owners switched to it consistently, and the removed `SAFE_NONCE_OVERRIDES` map was empty on main so nothing active is lost. The dynamic `getNextNonce()` matches the Safe Kit's prior default behavior.

My main ask is to **pull the `tryUpdateContainerIsm` change out of this PR** so the config fix isn't held up. This is an area I've looked at before, and in-place container ISM updates are meaningfully more involved than the implementation here. The duplicate same-type pairing bug below is the clearest example, and the spread of follow-on issues/fixes surfaced in the stacked refactor (#8905) reinforces that it deserves its own PR with proper tests (duplicate same-type mutable children, preserved-address preflight) rather than riding along with a config change.

The `isInitialized` guard is worth keeping in this PR. It's low-risk and useful. Only the comment's rationale is a bit off (see inline). The nonce change is also fine to keep here.

Suggested split:
- Keep here: BSC config, `isInitialized` guard, dynamic nonce.
- Move to #8905 (or its own PR): `tryUpdateContainerIsm` + the `update()` early in-place path.

Non-blocking nits inline. Not requesting changes.

File: typescript/sdk/src/ism/EvmIsmModule.ts [hunk 1]
```diff
@@ -493,4 +517,101 @@ export class EvmIsmModule extends HyperlaneModule<
...
+        const aType =
+          typeof a === 'object' && a !== null ? (a as any).type : String(a);
+        const bType =
+          typeof b === 'object' && b !== null ? (b as any).type : String(b);
+        return aType < bType ? -1 : aType > bType ? 1 : 0;
+      });
+      subModules = onChainTyped.map(({ addr }, i) => ({
+        address: addr,
+        targetConfig: sortedTargetModules[i],
+      }));
```
Comment (paulbalaji) [inline typescript/sdk/src/ism/EvmIsmModule.ts]: Pairing-by-sorted-type is unsafe when sub-modules share a type. On-chain modules are sorted by derived `.type` and target modules by config `.type`, then zipped by index. If two children have the same `IsmType` (e.g. two `MERKLE_ROOT_MULTISIG`, or two routing modules), the comparator returns 0 and ordering is ambiguous, so an on-chain address can bind to the wrong target config. The `eqAddress` guard at line 611 does **not** catch this: a mutable child intentionally keeps its address after an in-place update, so a wrong-child mutation passes the check and silently misconfigures the aggregation.

Not triggered by the Citrea/Moonpay routes (their aggregations are `[AMOUNT_ROUTING, FALLBACK_ROUTING]`, distinct types), but it's a latent landmine for any future aggregation with repeated types. Minimum fix: bail to redeploy (`return null`) when duplicate types are present. Better handled in the dedicated PR (#8905) with tests. Also `(a as any).type` / `String(a)` here is dead + a cast violation: at this point target modules are already fully derived objects with `.type`, and `normalizeConfig` already sorts modules by `.type`, so only the on-chain side needs sorting.

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts [hunk 2]
```diff
@@ -640,16 +641,29 @@ export class HyperlaneIsmFactory extends HyperlaneApp<ProxyFactoryFactories> {
...
-        receipt = await this.multiProvider.handleTx(
-          destination,
-          routingIsm['initialize(address,uint32[],address[])'](
-            config.owner,
-            safeConfigDomains,
-            submoduleAddresses,
-            overrides,
-          ),
-        );
+        // Guard against re-initialization when a previous run deployed this
```
Comment (paulbalaji) [inline typescript/sdk/src/ism/HyperlaneIsmFactory.ts]: Worth keeping the guard, but the rationale is slightly misleading. The fallback routing ISM is deployed via a fresh `handleDeploy` (new CREATE address each run), so on a re-run after a mid-deploy crash this lands on a brand-new uninitialized address and the guard never fires. So it doesn't actually make the fallback path idempotent across runs (the ZkSync path with a deterministic address is where it can hit). Suggest softening the comment to reflect that it's a defensive double-init guard rather than crash-recovery. The detection itself is correct (reads slot 0, OZ 4.9.6 `_initialized` == 1 or 255).

File: typescript/infra/src/govern/multisend.ts [hunk 3]
```diff
@@ -104,19 +100,26 @@ export class SafeMultiSend extends MultiSend {
     }
   }
 
+  private async getNextNonce(): Promise<number> {
+    const nextNonce = await retrySafeApi(() =>
+      this.safeService.getNextNonce(this.safeAddress),
+    );
+    return parseInt(nextNonce);
```
Comment (paulbalaji) [inline typescript/infra/src/govern/multisend.ts]: Nit: add explicit radix. `safeService.getNextNonce` returns a `Promise<string>`, so `parseInt` is the right call, but `parseInt(nextNonce, 10)` matches the convention already used in `safe.ts`.

File: .changeset/hot-bears-enjoy.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+"@hyperlane-xyz/infra": patch
+"@hyperlane-xyz/sdk": patch
+---
+
+Fixed bsc owner and owner updates in sdk
```
Comment (paulbalaji) [inline .changeset/hot-bears-enjoy.md]: Nit: a bit vague. Once the SDK change is split out, this becomes infra-only and could read closer to: "Migrated Moonpay BSC USDC/USDT warp route owners to the BSC ICA and replaced the static Safe nonce override map with dynamic next-nonce fetching."

File: typescript/sdk/src/ism/EvmIsmModule.ts
Comment (troykessler) [reply typescript/sdk/src/ism/EvmIsmModule.ts:1]: moved the bsc owner updates to https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/8885 and this pr now focuses only on the sdk fixes

File: typescript/infra/src/govern/multisend.ts [hunk 5]
```diff
@@ -19,10 +19,6 @@ import {
   retrySafeApi,
 } from '../utils/safe.js';
 
-// Safe nonce overrides to ensure transactions are proposed at the correct nonce.
-// Remove entries once the transactions have been executed.
-export const SAFE_NONCE_OVERRIDES: Record<string, number> = {};
```
Comment (paulbalaji) [inline typescript/infra/src/govern/multisend.ts:24]: no we still want this because it's useful for manually submitting at specific nonces

which is why we had that resolveBaseNonce function too

File: typescript/infra/src/govern/multisend.ts [hunk 6]
```diff
@@ -104,26 +100,18 @@ export class SafeMultiSend extends MultiSend {
...
-  // Resolve the base nonce for new proposals: a manual override if set,
-  // otherwise the Safe transaction service's next nonce (queue-aware: highest
-  // pending + 1). Falling back to protocol-kit's default would use the on-chain
-  // nonce and collide with an already-pending proposal at that nonce.
-  private async resolveBaseNonce(): Promise<number> {
-    const override = SAFE_NONCE_OVERRIDES[this.chain];
-    if (override !== undefined) {
-      return override;
-    }
+  private async getNextNonce(): Promise<number> {
```
Comment (paulbalaji) [inline typescript/infra/src/govern/multisend.ts]: Let's keep `SAFE_NONCE_OVERRIDES` and `resolveBaseNonce` rather than removing them. As noted in #8895, the override map is intentional — it's how we manually propose at a specific nonce. `getNextNonce()` is the right *fallback*, so the shape should stay: override if set, else the queue-aware next nonce.

This removal also breaks the infra build — `typescript/infra/test/multisend.test.ts:11` still imports `SAFE_NONCE_OVERRIDES`, so `infra-test` fails with `TS2305: ... has no exported member 'SAFE_NONCE_OVERRIDES'`. Restoring the map fixes both. (Keeping the explicit `parseInt(nextNonce, 10)` radix is a nice touch.)

File: typescript/sdk/src/ism/EvmIsmModule.ts [hunk 7]
```diff
@@ -493,4 +517,103 @@ export class EvmIsmModule extends HyperlaneModule<
       mailbox: this.mailbox,
     });
   }
+
+  // Attempts to update AGGREGATION or AMOUNT_ROUTING sub-modules in-place.
+  // Returns accumulated transactions if all sub-module addresses are unchanged
+  // (container address preserved), or null to fall back to full redeployment.
+  private async tryUpdateContainerIsm(
```
Comment (paulbalaji) [inline typescript/sdk/src/ism/EvmIsmModule.ts:524]: Let's replace this with #8905's implementation and bring its hardhat tests over, then close #8905. The version here handles the happy path for the Citrea/Moonpay routes but misses the edge cases #8905 covers: ambiguous duplicate sort keys checked on *both* the on-chain and target sides (this version only dedupes target types and zips by index, which can mispair if the derived on-chain types don't line up after sort), nested containers, CCIP cache, and nested `RATE_LIMITED` recipient changes. Better to land the complete version once than iterate on the partial one.

File: typescript/sdk/src/ism/EvmIsmModule.ts [hunk 8]
```diff
@@ -493,4 +517,103 @@ export class EvmIsmModule extends HyperlaneModule<
...
+        this.multiProvider,
+        {
+          chain: this.chain,
+          config: targetConfig,
+          addresses: { ...this.args.addresses, deployedIsm: origAddress },
+        },
+        undefined,
+        this.contractVerifier,
+      );
+      allUpdateTxs.push(...(await subModule.update(targetConfig)));
```
Comment (paulbalaji) [inline typescript/sdk/src/ism/EvmIsmModule.ts]: `subModule.update()` calls `deploy()`, which deploys on-chain *immediately* — so when a child's address can't be preserved (e.g. a nested `RATE_LIMITED` recipient change, or a nested container needing redeploy), we deploy a replacement child here, then return `null` at L613 and fully redeploy the parent anyway, orphaning what we just deployed (wasted gas, never wired up). #8905 avoids this by deciding in a side-effect-free preflight *before* any deploy runs.

File: typescript/sdk/src/ism/EvmIsmModule.ts [hunk 9]
```diff
@@ -493,4 +517,103 @@ export class EvmIsmModule extends HyperlaneModule<
...
+    const allUpdateTxs: AnnotatedEV5Transaction[] = [];
+    for (const { address: origAddress, targetConfig } of subModules) {
+      const subModule = new EvmIsmModule(
+        this.multiProvider,
+        {
+          chain: this.chain,
+          config: targetConfig,
+          addresses: { ...this.args.addresses, deployedIsm: origAddress },
+        },
+        undefined,
```
Comment (paulbalaji) [inline typescript/sdk/src/ism/EvmIsmModule.ts]: Passing `undefined` here drops the CCIP contract cache for nested child updates. #8905 threads `this.ccipContractCache` through so the cache is preserved across the tree.

File: typescript/sdk/src/ism/EvmIsmModule.ts [hunk 10]
```diff
@@ -493,4 +517,103 @@ export class EvmIsmModule extends HyperlaneModule<
...
+  ): Promise<AnnotatedEV5Transaction[] | null> {
+    type SubModuleEntry = { address: Address; targetConfig: IsmConfig };
+    let subModules: SubModuleEntry[];
+    const provider = this.multiProvider.getProvider(this.chain);
+
+    if (
+      current.type === IsmType.AGGREGATION &&
+      target.type === IsmType.AGGREGATION
+    ) {
+      const currentAgg = current as AggregationIsmConfig;
```
Comment (paulbalaji) [inline typescript/sdk/src/ism/EvmIsmModule.ts]: Nit: these `as AggregationIsmConfig` / `as AmountRoutingIsmConfig` casts (and the `as { type: string }` further down) go against our zero-cast guideline. #8905 narrows via a shared sort-key helper instead — this comes over for free with that change.

**Review by paulbalaji — COMMENTED** [review]
Thanks @troykessler! Reviewed alongside #8905 and #8895. Fix #1 (the `isInitialized` guard) and the direction of fix #3 (queue-aware next nonce) are both sound. A few things to sort before this lands.

**Plan for the container-ISM work:** let's consolidate rather than ship the partial version. Fold #8905's full implementation (side-effect-free preflight before any mutation, duplicate-key checks on *both* the on-chain and target sides, CCIP cache propagation, nested `RATE_LIMITED` recipient handling) **and its hardhat tests** into this PR, then close #8905. That way the feature lands complete in one place. Inline notes mark the specific gaps.

**Blocker — keep `SAFE_NONCE_OVERRIDES`**
We deliberately kept this in #8895 (it's how we manually propose at a specific nonce, which is why `resolveBaseNonce` existed) — `getNextNonce()` should be the *fallback*, not a replacement. Removing the export also breaks the infra build: `typescript/infra/test/multisend.test.ts:11` still imports it, so `infra-test` fails with TS2305. Restore the map + `resolveBaseNonce` (override → else next nonce) and the test stays green. Inline at the function.

**Needs the #8905 treatment** (inline)
- Sub-modules are mutated sequentially before confirming all children preserve their addresses → can deploy orphaned contracts on-chain before falling back to a parent redeploy.
- Nested child modules are constructed with `undefined` ccipContractCache, dropping the cache across the tree.
- No `EvmIsmModule.hardhat-test.ts` coverage for the new path; #8905 already has the right set (duplicate-type fallback/non-mutation, nested containers, preflight-abort-without-side-effects, amount-routing fallback, nested `RATE_LIMITED`).

**Nit**
- A couple of `as` casts in `tryUpdateContainerIsm` (against our zero-cast guideline); #8905 removes these via a sort-key helper.

Not requesting changes — happy to approve once #8905's impl + tests are folded in here and the nonce override is restored.

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts (hunk 2 above)
Comment (troykessler) [reply typescript/sdk/src/ism/HyperlaneIsmFactory.ts]: Done — comment updated to say "Defensive double-init guard (fresh CREATE address, so this never fires in practice; kept for safety)."

File: typescript/infra/src/govern/multisend.ts (hunk 3 above)
Comment (troykessler) [reply typescript/infra/src/govern/multisend.ts]: Done — already using `parseInt(nextNonce, 10)`.

**Review by xeno097 — APPROVED** [review]
